### PR TITLE
Feature/12 heading s ru kf

### LINKED ISSUE #15: Heading-only SRuKF module
Cpp implementation of C SRuKF filter in order to perform heading only (unit vector input measurement) orbit determination.
Measurements come from the Center of Brightness converter module

### PR DESCRIPTION
**Pull request by thibaudteil**
* **Tickets addressed:** bsk-012
* **Review:** By commit 
* **Merge strategy:** Merge (no squash) 

## Description
This PR contains the first Cpp Square-Root unscented Kalman Filter (SRuKF) in Basilisk. It performs heading based orbit determination (using unit vector measurements). It should provide a baseline for future filter refactors.

## Verification
Module was tested in the unit test, and is being separately run in an integrated test.

## Documentation
Code was commented thoroughly and rst documentation was added

## Future work
Filter refactor in the future for better composability

### CODE REVIEW

File: src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h
Comment (patkenneally) [inline src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h:4]: :small_red_triangle_down: I didn't think of it until looking at this line specifically, but we need to change this line to ` Copyright (c) 2023, Laboratory  for Atmospheric and Space Physics, University of Colorado at Boulder`

When a new file is added to the repo this license line should be added to reflect the author org and year.

File: src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h
Comment (patkenneally) [inline src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h:30]: :large_blue_circle: I would align the comments one space against their respective code. The whole aligning them on the right is a losing battle. Like trying to dig a hole to close to the water on the beach.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: :large_orange_diamond: Some rogue comment control characters.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: :large_orange_diamond: Is it expected that these will change to lower case e.g. `timeUpdate(..` in some later refactor?

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h:56]: So much `const`! Nice.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: :large_blue_circle: Worth colocating this function definition just below `MeasurementUpdate()`?

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h:78]: :large_blue_circle: These variables aren't understandable from name and context. Should we add a comment?

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: :large_orange_diamond: Is this a thing or does it do a thing? Does it make more sense to be named `propagate` ?

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: :small_red_triangle_down: license change

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :small_red_triangle_down: license change

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.i
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.i]: :small_red_triangle_down: license change

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :small_red_triangle_down: This heading commentary doesn't match the parameters taken by `Reset`

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :large_orange_diamond: This could be factored out into a `readMessage` method.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :large_orange_diamond: This could be factored out into a `computePostFitResiduals` method.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :large_orange_diamond: This could be factored out into a `writeMessages` method.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :large_orange_diamond: inline loop counters. This is a good safe practice as it limits the scope of the variable and allows the compiler to take advantage of potential optimizations.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :large_blue_circle: You could declare/assign this in one line if you like.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp:257]: :large_blue_circle: You could declare/assign this in one line if you like.

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: :large_blue_circle: You could drop each parameter onto a new line like
```
Eigen::VectorXd FlybyODuKF::RK4(const std::function<Eigen::VectorXd(double, Eigen::VectorXd)>& ODEfunction,
                                const Eigen::VectorXd& X0, 
                                double t0, 
                                double dt) const
```

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: 🔵 You could drop each parameter onto a new line like

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/_UnitTest/flybyODuKF_test_utilities.py
Comment (patkenneally) [inline src/fswAlgorithms/opticalNavigation/flybyODuKF/_UnitTest/flybyODuKF_test_utilities.py]: :small_red_triangle_down: File ending new line

**Review by patkenneally — APPROVED** [review]
Legend:
:small_red_triangle_down: *Issues* to address before merge
:large_orange_diamond: *Requests* that should not block merge, but should at least be discussed
:large_blue_circle: *Recommendations* that can be ignored if desired

File: src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h
Comment (thibaudteil) [reply src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h:4]: Done

File: src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h
Comment (thibaudteil) [reply src/architecture/msgPayloadDefCpp/OpNavSUKFMsgPayload.h:30]: Will do

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: Fixed

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h:78]: comment added

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: it does make sense to call it propagate!

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.h]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.i
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.i]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Thanks fixed

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/_UnitTest/flybyODuKF_test_utilities.py
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/_UnitTest/flybyODuKF_test_utilities.py]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp:257]: Done

File: src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp
Comment (thibaudteil) [reply src/fswAlgorithms/opticalNavigation/flybyODuKF/flybyODuKF.cpp]: Done

**Pull request merged**
